aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -165,67 +165,64 @@ void cpu_upsample_nearest_channels_last(
       int64_t iw = nearest_idx(ow, input_width, output_width, scales[1]);
-      auto input_data_ptr = input_data + n * (input_height * input_width * channels)
-        + ih * (input_width * channels) + iw * channels;
-      auto output_data_ptr = output_data + n * (output_height * output_width * channels)
-        + oh * (output_width * channels) + ow * channels;
-      std::memcpy(output_data_ptr, input_data_ptr, sizeof(scalar_t) * channels);
+      scalar_t* output_ptr = output_data + i * channels;
+      scalar_t* input_ptr = input_data + n * input_height * input_width * channels +
+          ih * input_width * channels + iw * channels;
+      copy(output_ptr, input_ptr, channels);
       data_index_step(n, num_batches, oh, output_height, ow, output_width);
     }
   };
 
-  auto loop3d = [&](int64_t start, int64_t end) {
+  auto loop3d = [&](int64_t begin, int64_t end) {
     int64_t n = 0;
     int64_t od = 0;
     int64_t oh = 0;
     int64_t ow = 0;
-    data_index_init(start, n, num_batches, oh, output_height, ow, output_width);
-    data_index_init(start, n, num_batches, od, output_depth, oh, output_height, ow, output_width);
+    data_index_init(begin, n, num_batches, od, output_depth, oh, output_height, ow, output_width);
 
-    // We need to unroll this loop by some factor. Perhaps by 4 or something.
-    // But we are doing memcpy then it may not help.
-    // Then we need custom sort of memcpy.
-    for (int64_t i = start; i < end; i++) {
+    for (int64_t i = begin; i < end; i++) {
       int64_t id = nearest_idx(od, input_depth, output_depth, scales[0]);
       int64_t ih = nearest_idx(oh, input_height, output_height, scales[1]);
       int64_t iw = nearest_idx(ow, input_width, output_width, scales[2]);
-      auto input_data_ptr = input_data
-        + n * (input_depth * input_height * input_width * channels)
-        + id * (input_height * input_width * channels)
-        + ih * (input_width * channels) + iw * channels;
-      auto output_data_ptr = output_data
-        + n * (output_depth * output_height * output_width * channels)
-        + od * (output_height * output_width * channels)
-        + oh * (output_width * channels) + ow * channels;
-      std::memcpy(output_data_ptr, input_data_ptr, sizeof(scalar_t) * channels);
+      scalar_t* output_ptr = output_data + i * channels;
+      scalar_t* input_ptr = input_data + n * input_depth * input_height * input_width * channels +
+          id * input_height * input_width * channels +
+          ih * input_width * channels + iw * channels;
+      copy(output_ptr, input_ptr, channels);
       data_index_step(n, num_batches, od, output_depth, oh, output_height, ow, output_width);
     }
   };
 
   if (ndim == 4) {
     // upsample nearest 2d
-    at::parallel_for(0, num_work_items, at::internal::GRAIN_SIZE, loop2d);
+    at::parallel_for(0, numel / channels, at::internal::GRAIN_SIZE / channels, loop2d);
```

Comment:
Do shape check make sure that channels is not 0? Otherwise the code will crash here

Reply:
add check for `channels`.